Contributing valkeyJSON module

roshkhatri opened the pull request:
This PR adds the implementation of valkeyJSON module based on the specifications detailed in the accepted RFC document: [ValkeyJSON Specification](https://github.com/valkey-io/valkey-rfc/blob/main/ValkeyJSON.md).

The implementation does not include JSON.MSET implementation but I am working on it as we review this. I am hoping we could add the implementation post merging this PR.

@valkey-io/core-team

madolson commented on src/json/json_api.h:
```diff
@@ -0,0 +1,70 @@
+/**
+ * JSON C API for Search Module
```
What's a search module?

joehu21 replied:
Let's just say "JSON C API".

hwware commented:
Sorry，what's the blocker for  JSON.MSET implementation? ^_^?

roshkhatri commented:
> Sorry，what's the blocker for JSON.MSET implementation? ^_^?

There is no blocker. its WIP. That can be added in a separate PR.

zuiderkwast commented on CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+# Define valkey-bloom branch
+set(VALKEY_BLOOM_BRANCH "unstable" CACHE STRING "Valkey-bloom branch to use")
+
+# Set the download directory for Valkey-bloom
+set(VALKEY_BLOOM_DOWNLOAD_DIR "${CMAKE_CURRENT_BINARY_DIR}/_deps/valkey-bloom-src")
+
+# Download valkey-bloom
+ExternalProject_Add(
+    valkey-bloom
+    GIT_REPOSITORY https://github.com/valkey-io/valkey-bloom.git
```
I would like to see some way to not download dependencies during the build, at least optionally.

Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the main concern. License/legal problems is another.

I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

zuiderkwast replied:
> I would like to see some way to not download dependencies during the build, at least optionally.
> 
> Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the main concern. (License issues is another concern.)
> 
> I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

PingXie replied:
looks like this is all about reusing the tests at https://github.com/valkey-io/valkey-bloom/tree/unstable/tests? or there is more?

joehu21 replied:
Yes, it's all about reusing the pytest infrastructure in valkey-bloom.

madolson replied:
https://github.com/valkey-io/valkeyJSON/issues/10

zuiderkwast reviewed (approved):
There is no way that I can review all this. 27K in one commit. :D

There will need to be maintainer team for each module that can review each other's PRs. So, at least two maintainers would be useful. Is there anyone besides you @roshkhatri?

I'd say let's merge it! Later, we can talk about beta releases and stable releases. The TSC can oversee your release plans and the work on a high level, but I don't see that TSC members have enough time to be deeply involved in these modules.

roshkhatri commented:
>There will need to be maintainer team for each module that can review each other's PRs. So, at least two maintainers would be useful. Is there anyone besides you @roshkhatri?

Yes, @joehu21  will be the maintainer besides me.

>I'd say let's merge it! Later, we can talk about beta releases and stable releases. The TSC can oversee your release plans and the work on a high level, but I don't see that TSC members have enough time to be deeply involved in these modules.

Yes, once this is merged we can open issues and discuss/work on them.

hwware commented:
Mostly it looks like fine, agree to merge it. Great work.

joehu21 reviewed (approved):
LGTM

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+    return handleHashTableFactor(&rapidjson::HashTableFactors::minHTSize, &val, size_t(1));
+}
+
+int registerModuleConfigs(ValkeyModuleCtx *ctx) {
```
The RFC has 2 configs, why does this have like 50?

zuiderkwast replied:
5️⃣0️⃣⁉️

It's easy to add configs but it's a breaking change to remove one. Without looking at what they are for, can we replace them with constants or something before we release a version where we commit to maintaining them?

joehu21 replied:
We considered dropping these configs. However, it requires a decent amount of code cleanup. Agreed we should remove them. We can create an issue to replace these configs with constants.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/2

madolson commented on src/json/stats.h:
```diff
@@ -0,0 +1,138 @@
+/**
+ * The STATS module's main responsibility is to produce the following metrics:
+ * 1. Core metrics:
+ *    json_total_memory_bytes: total memory allocated to JSON objects
```
The original rfc has 2 metrics, why does this one have so much more?

joehu21 replied:
These module infos have been removed (not emitted). However, the documentation is forgotten to cleaned up.

madolson replied:
https://github.com/valkey-io/valkeyJSON/issues/11

madolson reviewed (commented):
Just evaluated the RFC against the implementation. There seems to be some discrepancies, so would like to start conversations about why.

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+ * Check the encoding version, For unsupported versions we ALWAYS put out a message in the log
+ * but we only fail the RDB load if the config tells us to do it.
+ */
+bool checkVersion(const char *type_name, int encver, int expected_encver) {
```
I suppose I don't get what this is for. What does this do for compatibility?

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+ * Check the encoding version, For unsupported versions we ALWAYS put out a message in the log
+ * but we only fail the RDB load if the config tells us to do it.
+ */
+bool checkVersion(const char *type_name, int encver, int expected_encver) {
```
I suppose I don't get what this is for. What does this do for compatibility?

joehu21 replied:
The function is invoked when loading a RDB produced by RedisJSON. Their RDB could include a newer version of a module type. This function decides whether or not to reject the RDB.

madolson replied:
Won't it just get rejected if it's from a newer version?

joehu21 replied:
Default value of enforce_rdb_version_check is false. So, the check will pass. In our internal RDB compatibility tests, which isn't part of the PR, we set enforce_rdb_version_check to true so that we will be alerted when ReJSON introduced some new aux types. In retrospect, I think the aux type version checking is an overkill. We should just read and ignore these aux types.

madolson replied:
https://github.com/valkey-io/valkeyJSON/issues/12

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+}
+
+/*
+ * Check an integer value, fail
```
Overall, this code is not very well documented. This comment just trails off for example.

joehu21 replied:
Yeah, the comment should be "Check an integer value, fail if the value doesn't match the expected one."

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+        return VALKEYMODULE_ERR;
+    }
+
+    /*
+     * Now create the stub datatypes for search
+     */
+    if (!install_stub(ctx, "scdtype00", SCDTYPE_ENCVER, scdtype_aux_load)) return VALKEYMODULE_ERR;
+    if (!install_stub(ctx, "GEARS_DT0", GEARSDT_ENCVER, gearsdt_aux_load)) return VALKEYMODULE_ERR;
+    if (!install_stub(ctx, "GEAR_REQ0", GEARSRQ_ENCVER, gearsrq_aux_load)) return VALKEYMODULE_ERR;
```
Apparently we also load 3 custom module types here. I understand this to be for compatibility with stack, but do we want to have this in this module?

zuiderkwast replied:
Interesting. At least, this should be added to the rfc.

joehu21 replied:
Yes, that's for RDB compatibility with RedisJSON. If we don't care about them any more, we can remove the code.

zuiderkwast replied:
I think we do care about compatibility with RedisJSON. We want to make it easy to migrate to Valkey.

I just wish it is mentioned in the RFC, which already has a lot of details.

madolson replied:
These aren't for json compatibility though, they list gears and search types.

zuiderkwast replied:
Which are partially used by the JSON module or vice versa? Or the JSON module provides a minimal subset of them or something? I don't know this. :laughing:

I guess the only reason to add them is to support RDB streams containing module types from these modules. Otherwise, delete it.

PingXie replied:
yeah these aux data types seem to just play a supporting role so far. can we register a generic stub instead and just log a message that "xxx is skipped" and move on? there could be more aux types added in the future in rejson so it will be good to have that visibility if community members report any issue.

```c
install_stub(ctx, "*", GENERIC_ENCVER, generic_aux_load);
```

joehu21 replied:
> These aren't for json compatibility though, they list gears and search types.

When testing compatibility with ReJSON, we use docker image "docker run -p 6389:6379 redis/redis-stack:latest", which loads a bunch of modules including search, graph, etc. Although our test only creates json keys, the generated RDB contains a bunch of aux types like scdtype00, gear*, etc. I suspect these aux types are from RedisSearch module, but not 100% sure because I never looked at their source code. (NOTE: It also indicates that their modules do not use aux_save2, which would not save aux types that have no corresponding data.)

To be able to load the RDB, we register these aux types, read the data and ignored them. So, these are indeed for JSON compatibility test.

PingXie replied:
You are right @joehu21. These aux data types are indeed needed. The main purpose of these handlers is actually to skip these aux types.

zuiderkwast replied:
If we ever want another module that provides compatibilty with redis search or gears modules, then these names are already registered by ValkeyJSON module so the other modules can't be loaded? That's a potential future problem.

joehu21 replied:
That's a good point.

madolson replied:
Yeah, I think we need a better overall strategy here instead of install random types and ignoring the data from them.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/9

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+        ValkeyModule_Log(ctx, "warning", "Failed to create command JSON.FORGET.");
+        return VALKEYMODULE_ERR;
+    }
+    if (ValkeyModule_SetCommandACLCategories(ValkeyModule_GetCommand(ctx,"JSON.FORGET"), cat_fast_write) == VALKEYMODULE_ERR) {
```
This code doesn't look linted, maybe create a followup issue to make sure we actually running the linter automatically. I guess generally we're missing workflow actions.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/3

roshkhatri replied:
Added the CI to run all the tests for now.

madolson reviewed (commented):
I am mostly trusting the code, can we setup github workflows to make sure this code all builds and what not?

madolson commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+ * 4. register commands that are all prefixed with "JSON.".
+ *
+ * Design Considerations:
+ * 1. Command API: see API.md.
```
👻 Mystery file in API.md.

joehu21 replied:
The line should be removed.

madolson commented:
Ideally let's make issues for the comments here. The only thing I think we should fix before merging is getting the CI working.

roshkhatri commented:
I have added the CI workflow to run against valkey `unstable` and `8.0.0` for now.

![Screenshot 2024-11-26 at 12 57 52 PM](https://github.com/user-attachments/assets/b31374c3-4cb9-4ed3-8c82-f79c98c43a5c)

PingXie commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,25 @@
+      matrix:
+        server_version: ['unstable', '8.0.0']
+    steps:
+    - uses: actions/checkout@v4
```
consider using the commit hash for the action instead.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/4

PingXie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+elseif("${ARCHITECTURE}" STREQUAL "aarch64")
+    message("Building JSON for aarch64")
+else()
+    message(FATAL_ERROR "Unsupported architecture. JSON is only supported on x86_64 and aarch64.")
```
print out the architecture?

joehu21 replied:
Yes, printing out `${ARCHITECTURE}` will be helpful.

joehu21 replied:
> looks like this is all about reusing the tests at https://github.com/valkey-io/valkey-bloom/tree/unstable/tests? or there is more?

Yes, it's for reusing the pytest infrastructure in valkey-bloom.

PingXie replied:
hmm this dependency is a bit odd - does it make sense to pull the pytest infra into the Valkey repo instead? We are discussing migrating to python based integration test infra for Valkey too.

roshkhatri replied:
Yes, I would suggest we should pull it out in the valkey repo instead. However, we can change the line later when we have it separately.

madolson replied:
I was going to suggest we write an RFC or something similar for the python testing framework as well before we get too far along building it organically.

zuiderkwast replied:
Why not move the testing framework to it's own repo to start with? It only needs to be pulled in for running the tests, not when just building the module.

I'd rather see people spend time on a README for the test framework than on an RFC. Or we can do both in parallel...?

madolson replied:
> I'd rather see people spend time on a README for the test framework than on an RFC. Or we can do both in parallel...?

I just want us to have some long term plan for testing and not just build it up organically. The reason is that Amazon internally has a similar testing framework that is extremely complex because nobody really maintains it and just bolt on random features. There are many cases where it's possible to do the same way 2 or 3 different ways. I think the same is true of our TCL framework, just to a smaller scale. 

So whether it be an RFC or just some direction in a new repo, I don't care which.

PingXie commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,25 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        server_version: ['unstable', '8.0.0']
```
should we go with the tip of the release branch, which is 8.0? Also I think having older engine versions (like 7.2) at some point would help catch regressions. worst case though we could end up with 6 versions based on our support promise.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/5

PingXie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+# Always include debug symbols and optimize the code
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O2 -g")
```
not going with -O3?

joehu21 replied:
We want the max optimization that still gives us debug symbols with line numbers in the crash stack. I believe such opt level is -O2.

PingXie replied:
maybe at least make is configurable on the command line so it could be overwritten if so desired?

joehu21 replied:
agreed. it's desirable to be able to override it through command line argument. @roshkhatri can you create an issue after the PR is merged?

madolson replied:
There are still debug symbols with a crash at -O3, you might be thinking about frame pointers which allow unwinding the stack. You can keep those with `-fno-omit-frame-pointer`.

> @roshkhatri can you create an issue after the PR is merged?

Please create the issues before the PR is merged. It's better practice to define the follow up work before merging in the current code.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/6

PingXie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+elseif("${ARCHITECTURE}" STREQUAL "aarch64")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=armv8-a")
+else()
+    message(FATAL_ERROR "Unsupported architecture. JSON is only supported on x86_64 and aarch64.")
```
nit - print out the actual arch too?

PingXie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+# Add the src subdirectory for building
+add_subdirectory(src)
+
+# Add the src subdirectory for building
```
btw, this comment can be skipped IMO
```suggestion
# Add the tst subdirectory for building
```

PingXie commented on README.md:
```diff
@@ -0,0 +1,89 @@
+ValkeyJSON leverages [RapidJSON](https://rapidjson.org/), a high-performance JSON parser and generator for C++, chosen for its small footprint and exceptional performance and memory efficiency. As a header-only library with no external dependencies, RapidJSON provides robust Unicode support while maintaining a compact memory profile of just 16 bytes per JSON value on most 32/64-bit machines.
+
+## Motivation
+While Valkey core lacks native JSON support, there's significant community demand for JSON capabilities. ValkeyJSON provides a comprehensive open-source solution with extensive JSON manipulation features.
```
this motivation belongs more to the rfc. Developers who got here would probably care more about the how-tos IMO.

PingXie commented on build.sh:
```diff
@@ -0,0 +1,63 @@
+    export SERVER_VERSION="unstable"
+fi
+
+if [ "$SERVER_VERSION" != "unstable" ] && [ "$SERVER_VERSION" != "8.0.0" ] ; then
```
this version check seems too restrictive. this tool is developer facing so might just open it up to all versions/commits

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/7

PingXie commented on src/json/json.cc:
```diff
@@ -0,0 +1,3232 @@
+
+#define GEARSDT_ENCVER 3
+int gearsdt_aux_load(ValkeyModuleIO *ctx, int encver, int when) {
+    if (!checkVersion("gearsdt", encver, GEARSDT_ENCVER)) return VALKEYMODULE_ERR;
+    if (when == VALKEYMODULE_AUX_AFTER_RDB) {
+        if (!checkString(ctx, "StreamReader", "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkInt(ctx, 0, "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkString(ctx, "CommandReader", "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkInt(ctx, 0, "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkString(ctx, "KeysReader", "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkInt(ctx, 0, "gears-dt")) return VALKEYMODULE_ERR;
+        if (!checkString(ctx, "", "gears-dt")) return VALKEYMODULE_ERR;
+    }
```
these checks add fragility only to the module compat. Since we are not planning to consume any of them, the best strategy imo would be to log a "xxx is skipped" notice level message and return _OK. same goes to other stubs.

joehu21 replied:
> yeah these aux data types seem to just play a supporting role so far. can we register a generic stub instead and just log a message that "xxx is skipped" and move on? there could be more aux types added in the future in rejson so it will be good to have that visibility if community members report any issue.

The challenge is that each aux type may have different number of entries in RDB. For example, scdtype only has one integer, while gearsdt has 7 entries in RDB (some are int and some are string).

PingXie replied:
Got it - the parsing part makes sense to me now. But maybe we could skip the check of the actual value?

joehu21 replied:
Totally agreed, no need to check the actual value.

roshkhatri replied:
https://github.com/valkey-io/valkeyJSON/issues/8

PingXie reviewed (approved):
This does look overwhelming and I only reviewed haphazardly. I agree we should just merge this PR and polish it incrementally, including adding more test coverage.

Thanks @roshkhatri and @joehu21!

madolson commented:
Before we merge, please just make sure everything has issues so we can track them.

enjoy-binbin reviewed (approved):
did not review, but it is indeed a good work, thanks.

roshkhatri commented:
I have created relevant issues for the comments and resolved those conversations to track them better. I have also addressed some feedback for minor changes.

Let me know if any more issues are required to be opens before we merge this PR. Happy Thanksgiving Everyone!!

hwware commented:
> I have created relevant issues for the comments and resolved those conversations to track them better. I have also addressed some feedback for minor changes.
> 
> Let me know if any more issues are required to be opens before we merge this PR. Happy Thanksgiving Everyone!!

Happy Black Friday !

madolson commented:
We have 6/6 TSC actually approved, so going ahead and merging it.
